### ISSUE TITLE
drive: increase pacer to 100ms by default

### ISSUE THREAD

**Issue by ncw** [ncw/rclone#2880]
Rclone currently uses a minimum pacer time for the google drive backend of 10mS which equates to 100 transactions per second.

However the maximum allowed quota per user is 1000 transactions in 100 seconds.  Even rclone's key can't exceed this.

This means that we are potentially doing more work, receiving more 403 errors and increasing the chance of a ban.

I propose to reduce the minimum pacer timeout to 100mS which is equivalent to 10 transactions per second.

In support of this, here are some tests made with an otherwise unused client_id with the user quota set to 1000 transactions per 100s.  These show uploading 100 small fles and listing 2,400 files (using `rclone size`) each in a reasonably normal file system heirachy.

These support quite strongly the idea that tpslimit 10 is the correct limit.

![image](https://user-images.githubusercontent.com/536803/50549736-8e917c80-0c5a-11e9-9ac5-fb6360de8457.png)

![image](https://user-images.githubusercontent.com/536803/50549743-99e4a800-0c5a-11e9-84b7-88cd94419b8b.png)

![image](https://user-images.githubusercontent.com/536803/50549749-a7019700-0c5a-11e9-9693-72f2ad917118.png)

![image](https://user-images.githubusercontent.com/536803/50549754-b97bd080-0c5a-11e9-9558-e6a29c0545cb.png)

I haven't tested how equivalent exactly `--tps-limit` and the pacer limits are though.

So questions

1. Is this a good idea?
2. Should the pacer be configurable by the user?
  - For drive min-sleep is the only parameter but there are others for the other backends

Thoughts @animosity22 and @B4dM4n ?

I'll do a test implementation with the default pacer set to 100ms and making it configurable and post it here in a moment so we can experiment.

**Comment by ncw** [comment]
I've made a quick test here

https://beta.rclone.org/branch/v1.45-057-g02c84fce-fix-2880-drive-pacer-beta/ (uploaded in 15-30 mins)

This has the new flag `--drive-pacer-min-sleep 100ms` the default is `100ms`.

Experimenting with this and comparing it to `10ms` I see that if you are doing small operations that don't normally return a 403 then it is slower than `100ms` as soon as the 403's kick in then they are the same.

It looks like this threshold is about 100 transactions, so you can get ~100 transactions for "free" before the rate limiting kicks in.  So I wonder if we should attempt to model that.  This is something that https://godoc.org/golang.org/x/time/rate can already do so maybe we should be using that in the pacer.

**Comment by animosity22** [comment]
I think that's a great idea as having more of the defaults set proper for GD makes the 'out of the box' setup much easier. 

As I've seen, we have a lot of programs that seem to be popping up with odd bugs that break things due to how they are setup and how they request things. If we can prevent those issues, that helps everyone out.

**Comment by ajkis** [comment]
Great will test it after new year ( cant access server now and dont wanna risk something going wrong )

**Comment by calisro** [comment]
Looks good in the surface. I'll have a go. I have services that I've kept at the default (smaller batches) and ones I've raised tps for to 10 already. I'll see what impact this has. It's always nice to have things user configurable but that also means more questions and people setting it incorrectly.

**Comment by ajkis** [comment]
@ncw it seems one of the flags is broken in this version 
Error: invalid argument "128MB" for "--drive-chunk-size" flag: parsing "128MB" as fs.SizeSuffix failed: strconv.ParseFloat: parsing "128M": invalid syntax

**Comment by B4dM4n** [comment]
> @ncw it seems one of the flags is broken in this version
> 
> `Error: invalid argument "128MB" for "--drive-chunk-size" flag: parsing "128MB" as fs.SizeSuffix failed: strconv.ParseFloat: parsing "128M": invalid syntax`
> 

@ajkis The unit symbol in `fs.SizeSuffix` is only one char, like `2b, 128m, 5G`.

> Experimenting with this and comparing it to `10ms` I see that if you are doing small operations that don't normally return a 403 then it is slower than `100ms` as soon as the 403's kick in then they are the same.
> 
> It looks like this threshold is about 100 transactions, so you can get ~100 transactions for "free" before the rate limiting kicks in. So I wonder if we should attempt to model that. This is something that https://godoc.org/golang.org/x/time/rate can already do so maybe we should be using that in the pacer.

Being able to burst requests is probably a really good feature when using `rclone mount`, as a default sleep time of `100ms` would slow it down substantially when only using requests sporadically.

The https://godoc.org/golang.org/x/time/rate package looks really promising for this.

**Comment by animosity22** [comment]
@B4dM4n - I'm not quite sure how to test, but can you actually burst on the Google Drive or do you get limited for that as well? I wasn't quite sure how to test if you can do say 900 transactions in 1-2 seconds and none the next 88-89 to meet the 1000 per 100 seconds or if you just got limited as well so the bursting didn't work either.

**Comment by ajkis** [comment]
@ncw Uploads still getting stuck 

`rclone sync dropbox: gdrive: --retries=5 --transfers=10 --checkers=4  --tpslimit=2 --drive-pacer-min-sleep=100ms  --log-level=INFO --log-file=$LogFile`

Part of the log where it get stuck ( unfortunately log level is at info )
basically once error occurs all progress is just stuck at 0% for ages ( it would be great if rclone could recognize  that file transfer is stuck and just exit with error instead of keep going )

`Failed to copy: googleapi: Error 403: Rate Limit Exceeded, rateLimitExceeded`

```
2019/01/04 07:19:10 INFO  : 
Transferred:   	  717.454G / 880.476 GBytes, 81%, 10.894 MBytes/s, ETA 4h15m24s
Errors:                 4 (retrying may help)
Checks:             74272 / 74272, 100%
Transferred:           96 / 115, 83%
Elapsed time:  18h44m0.5s
Transferring:
 *     crypt/hm2ka8or5qp1peâ€¦hc60gkllnflfbnjikb24:  0% /9.729G, 0/s, -
 *     crypt/hm2ka8or5qp1peâ€¦ipmbjj7hik5jdgv3gbqk: 95% /10.568G, 5.402M/s, 1m38s

2019/01/04 07:20:03 ERROR : crypt/hm2ka8or5qp1pefhtc1e62jn50/lsuk349eq58mtg3u2rn0gchu5j1fd8ac4shgeo1q3qgbbk2kmvfg/cu4ueiaakh6hshlbcieu7g5d25bnbkink3ivts77uua077g99bqd1hq7dhc60gkllnflfbnjikb24: Failed to copy: googleapi: Error 403: Rate Limit Exceeded, rateLimitExceeded
2019/01/04 07:20:10 INFO  : 
Transferred:   	  717.790G / 870.747 GBytes, 82%, 10.889 MBytes/s, ETA 3h59m44s
Errors:                 5 (retrying may help)
Checks:             74272 / 74272, 100%
Transferred:           96 / 114, 84%
Elapsed time:  18h45m0.5s
Transferring:
 *     crypt/hm2ka8or5qp1peâ€¦cmli4e2henjc7j9h0dlu:  0% /9.448G, 0/s, -
 *     crypt/hm2ka8or5qp1peâ€¦ipmbjj7hik5jdgv3gbqk: 98% /10.568G, 5.897M/s, 32s

2019/01/04 07:20:45 INFO  : crypt/hm2ka8or5qp1pefhtc1e62jn50/khrpp0f6g2j7m2adsgcm4j4h2fdc21s9m81jme2aqh3ikn6c4nc0/crof01t515qmi1n3bjqtb2l7d36r16ctslaq91mfik6pfgil9f1h4m5rsipmbjj7hik5jdgv3gbqk: Copied (new)
2019/01/04 07:21:10 INFO  : 
Transferred:   	  717.975G / 870.747 GBytes, 82%, 10.882 MBytes/s, ETA 3h59m35s
Errors:                 5 (retrying may help)
Checks:             74272 / 74272, 100%
Transferred:           97 / 114, 85%
Elapsed time:  18h46m0.5s
Transferring:
 *     crypt/hm2ka8or5qp1peâ€¦5ovlko53js277ioqokgk:  0% /6.559G, 0/s, -
 *     crypt/hm2ka8or5qp1peâ€¦cmli4e2henjc7j9h0dlu:  0% /9.448G, 0/s, -

2019/01/04 07:22:10 INFO  : 
Transferred:   	  717.975G / 870.747 GBytes, 82%, 10.873 MBytes/s, ETA 3h59m48s
Errors:                 5 (retrying may help)
Checks:             74272 / 74272, 100%
Transferred:           97 / 114, 85%
Elapsed time:  18h47m0.5s
Transferring:
 *     crypt/hm2ka8or5qp1peâ€¦5ovlko53js277ioqokgk:  0% /6.559G, 0/s, -
 *     crypt/hm2ka8or5qp1peâ€¦cmli4e2henjc7j9h0dlu:  0% /9.448G, 0/s, -
```

**Comment by ajkis** [comment]
`rclone sync dropbox: gdrive: --retries=5 --transfers=10 --checkers=4  --tpslimit=2 --drive-pacer-min-sleep=100ms --bwlimit=35M --drive-chunk-size=128M --log-level=INFO --log-file=$LogFile`

Still getting 

`2019/01/06 08:46:32 ERROR : crypt/hm2ka8or5qp1pefhtc1e62jn50/05bcvetfol5rvlssrun8rlsacg: error reading destination directory: couldn't list directory: googleapi: Error 403: Rate Limit Exceeded, rateLimitExceeded`

**Comment by calisro** [comment]
What rclone version are you using? --version

**Comment by calisro** [comment]
Wonder if that is recreatable or a fluke.

**Comment by ajkis** [comment]
> Wonder if that is recreatable or a fluke.

Yea i wonder too. Planed to stop it and remove bwlimit but since its will working will just wait it out.
(my gdrive is still way behind the dropbox )

Still going strong 
```
2019/01/08 10:10:30 INFO  : 
Transferred:   	    5.225T / 5.619 TBytes, 93%, 30.522 MBytes/s, ETA 3h45m6s
Errors:                28 (retrying may help)
Checks:            223176 / 223176, 100%
Transferred:          194 / 216, 90%
Elapsed time:  49h52m0.4s
Transferring:
 *     crypt/0ibcad98vheukuâ€¦1niqjke6df4of3ap9oh0: 19% /50.926G, 3.510M/s, 3h18m44s
 *     crypt/0ibcad98vheukuâ€¦1v9bdn1loi4mfdt77ou2: 20% /48.405G, 3.476M/s, 3h9m41s
 *     crypt/0ibcad98vheukuâ€¦2dpfmbm5m77f5fmm7psk: 21% /56.085G, 3.481M/s, 3h35m58s
 *     crypt/0ibcad98vheukuâ€¦7lqjtei3kqgavdt2hkkg: 17% /69.469G, 3.546M/s, 4h35m12s
 *     crypt/0ibcad98vheukuâ€¦bku4sdjjnk61lubuelo8: 26% /47.144G, 3.484M/s, 2h50m12s
 *     crypt/0ibcad98vheukuâ€¦fdpc49eorf4ccmm80da8: 21% /52.072G, 3.502M/s, 3h19m55s
 *     crypt/0ibcad98vheukuâ€¦op0aap34r59ehtli4sgg: 27% /41.127G, 3.525M/s, 2h25m3s
 *     crypt/0ibcad98vheukuâ€¦rmh25clh54bq54316uqg: 21% /53.052G, 3.494M/s, 3h24m31s
 *     crypt/0ibcad98vheukuâ€¦tgfh873a192i7hboog08: 25% /48.363G, 3.520M/s, 2h55m15s
 *     crypt/hm2ka8or5qp1peâ€¦pfghhn1mgvulklkf8gi0: 80% /9.237G, 3.461M/s, 8m51s
```

**Comment by calisro** [comment]
I think before the default of 100ms is merged, we should consider the bursting.  It will have less impact on mounts and other infrequently used remotes if bursting was handled like @B4dM4n mentioned.  Thoughts?

**Comment by ncw** [comment]
https://beta.rclone.org/branch/v1.45-059-gf935acb2-fix-2880-drive-pacer-beta/ (uploaded in 15-30 mins)

This lets the pacer burst when it hasn’t used any calls recently - this should be what rclone mount needs.

The binary has a --driver-pacer-burst flag to control this.

**Comment by ncw** [comment]
I messed up the build, try

https://beta.rclone.org/branch/v1.45-103-gc9e50b32-fix-2880-drive-pacer-beta/ (uploaded in 15-30 mins)

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.46
